### PR TITLE
Show confirm dialog when clicking traces

### LINKED ISSUE #13714: In automations losing edits if you hit traces after you've forced a run
### The problem

The issue seems to be after making multiple edits in the latest automation editor and you hit the three buttons in the top right to run the automation and swiftly hit "traces" just to the left of that you'll lose all the changes you just added since you've last saved.

### What version of Home Assistant Core has the issue?

2022.9.1

### What was the last working version of Home Assistant Core?

_No response_

### What type of installation are you running?

Home Assistant OS

### Integration causing the issue

Automation

### Link to integration documentation on our website

https://www.home-assistant.io/docs/automation/

### Diagnostics information

N/A ? not sure 

### Example YAML snippet

```yaml
N/A
```

### Anything in the logs that might be useful for us?

```txt
no errors reported, just data lost.
```

### Additional information

Not sure if OS is the right type of install for the one I have on my raspberry pi.

### PR DESCRIPTION
**Pull request by piitaya**
## Proposed change

Show confirmation modal when user click on trace with unsaved changes

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #13714
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/config/automation/ha-automation-editor.ts [hunk 1]
```diff
@@ -486,6 +480,13 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
     fireEvent(this, "hass-more-info", { entityId: this._entityId });
   }
 
+  private async _showTrace() {
+    if (this._config?.id) {
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:484]: you should check for `_dirty`

File: src/panels/config/automation/ha-automation-editor.ts [hunk 2]
```diff
@@ -535,39 +536,32 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
...
-    } else {
-      afterNextRender(() => history.back());
-    }
-  };
-
-  private async _duplicate() {
-    if (this._dirty) {
-      if (
-        !(await showConfirmationDialog(this, {
+  private confirmUnsavedChanged() {
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:539]:
```suggestion
  private async confirmUnsavedChanged() {
```

File: src/panels/config/automation/ha-automation-editor.ts [hunk 3]
```diff
@@ -535,39 +536,32 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
...
+    return new Promise<void>((res) => {
+      if (this._dirty) {
+        showConfirmationDialog(this, {
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:542]:
```suggestion
        const result = await showConfirmationDialog(this, {
```

File: src/panels/config/automation/ha-automation-editor.ts [hunk 4]
```diff
@@ -535,39 +536,32 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
...
           ),
           confirmText: this.hass!.localize("ui.common.leave"),
           dismissText: this.hass!.localize("ui.common.stay"),
-        }))
-      ) {
-        return;
+          confirm: () => {
+            afterNextRender(() => res());
+          },
+        });
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:551]:
```suggestion
        });
```

File: src/panels/config/automation/ha-automation-editor.ts [hunk 5]
```diff
@@ -486,6 +480,13 @@ export class HaAutomationEditor extends KeyboardShortcutMixin(LitElement) {
     fireEvent(this, "hass-more-info", { entityId: this._entityId });
   }
 
+  private async _showTrace() {
+    if (this._config?.id) {
+      await this.confirmUnsavedChanged();
```
Comment (bramkragten) [inline src/panels/config/automation/ha-automation-editor.ts:485]: What if I dismiss this? Then I will still be navigated to traces and loose my changes...

Comment (bramkragten) [reply src/panels/config/automation/ha-automation-editor.ts:485]: Or we never resolve the promise then? That is not good... we should either reject it, or better just return a boolean

Comment (piitaya) [reply src/panels/config/automation/ha-automation-editor.ts:485]: Returning boolean is better I think 🙂

**Review by bramkragten — APPROVED** [review]

**Pull request merged**